src/libs/sfdk/dockermanager.cpp

```diff
@@ -0,0 +1,129 @@
+** conditions contained in a signed written agreement between you and
+** The Qt Company.
+**
+****************************************************************************/
+#include "dockermanager.h"
+
+#include "asynchronous_p.h"
+#include "sdk_p.h"
+
+#include <QDir>
+#include <QRegularExpression>
+
+namespace  {
+    const char DOCKER[] = "docker";
+    const char IMAGE[] = "image";
+    const char LIST[] = "ls";
+    const char FORMAT[] = "--format";
+    const char IMAGE_FORMAT[] = "{{.Repository}}";
+    const char UNNAMED_IMAGE[] = "<none>";
+}
+
+namespace Sfdk {
+
+DockerManager::DockerManager(QObject *parent)
+    : QObject(parent)
+{
+
+}
+
+DockerManager *DockerManager::s_instance = nullptr;
+
+DockerManager *DockerManager::instance()
+{
+    if (!s_instance) {
+        s_instance = new DockerManager();
+    }
+    return s_instance;
+}
+
+QStringList DockerManager::listedImages(const QString &output)
+{
+    QStringList images;
+    QStringList lines = output.split(QRegularExpression("[\r\n]"), QString::SkipEmptyParts);
+    for (QString line : lines) {
+        if (line != QLatin1String(UNNAMED_IMAGE))
+            images.append(line);
+    }
+    return images;
+}
+
+void DockerManager::fetchRegisteredVirtualMachines(const QObject *context, const Functor<const QStringList &, bool> &functor)
+{
+    Q_ASSERT(context);
+    Q_ASSERT(functor);
+
+    QStringList arguments;
+    arguments.append(QLatin1String(IMAGE));
+    arguments.append(QLatin1String(LIST));
+    arguments.append(QLatin1String(FORMAT));
+    arguments.append(QLatin1String(IMAGE_FORMAT));
+
+    auto runner = std::make_unique<ProcessRunner>(dockerPath(), arguments);
+    connect(runner.get(), &ProcessRunner::done,
+            context, [=, runner = runner.get()](bool ok) {
+        if (!ok) {
+            functor({}, false);
+            return;
+        }
+        QStringList vms = listedImages(QString::fromLocal8Bit(
+                runner->process()->readAllStandardOutput()));
+        functor(vms, true);
+    });
+    SdkPrivate::commandQueue()->enqueue(std::move(runner));
+}
+
+QStringList DockerManager::usedVirtualMachines()
+{
+    return instance()->m_usedImages.keys();
+}
+
+QString DockerManager::dockerPath()
+{
+    if (instance()->m_dockerPath.isNull()) {
+        QStringList searchPaths = QProcessEnvironment::systemEnvironment()
+            .value(QLatin1String("PATH")).split(QLatin1Char(':'));
+        foreach (const QString &searchPath, searchPaths) {
+            QDir dir(searchPath);
+            if (dir.exists(QLatin1String(DOCKER))) {
+                instance()->m_dockerPath = dir.absoluteFilePath(QLatin1String(DOCKER));
+                break;
+            }
+        }
+    }
+    return instance()->m_dockerPath;
+}
+
+void DockerManager::useImage(const QString &image)
+{
+    if (--(instance()->m_usedImages[image]) == 0) {
+        instance()->m_usedImages.remove(image);
+    }
```

Comment:
This is now enforced by VirtualMachineFactory and can be removed here. The other static functions should be moved to the DockerVirtualMachinePrivate class in a fashion similar to how it is done with VBoxVirtualMachinePrivate class.

Reply:
fixed